components/raftstore/src/store/worker/read.rs

```diff
@@ -165,13 +174,128 @@ pub struct ReadDelegate {
     pub track_ver: TrackVer,
 }
 
+/// CachedReadDelegate is a wrapper the ReadDelegate and kv_engine. LocalReader
+/// dispatch local read requests to ReadDeleage according to the region_id where
+/// ReadDelegate needs kv_engine to read data or fetch snapshot.
+pub struct CachedReadDelegate<E>
+where
+    E: KvEngine,
+{
+    delegate: Arc<ReadDelegate>,
+    kv_engine: E,
+}
+
+impl<E> Deref for CachedReadDelegate<E>
+where
+    E: KvEngine,
+{
+    type Target = ReadDelegate;
+
+    fn deref(&self) -> &Self::Target {
+        self.delegate.as_ref()
+    }
+}
+
+impl<E> Clone for CachedReadDelegate<E>
+where
+    E: KvEngine,
+{
+    fn clone(&self) -> Self {
+        CachedReadDelegate {
+            delegate: Arc::clone(&self.delegate),
+            kv_engine: self.kv_engine.clone(),
+        }
+    }
+}
+
+/// #[RaftstoreCommon]: LocalReadContext combines some LocalReader's fields for temporary usage.
+pub struct LocalReadContext<'a, E>
+where
+    E: KvEngine,
+{
+    metrics: &'a mut ReadMetrics,
+    read_id: &'a mut ThreadReadId,
+    snap_cache: &'a mut Box<Option<Arc<E::Snapshot>>>,
+}
+
 impl Drop for ReadDelegate {
     fn drop(&mut self) {
         // call `inc` to notify the source `ReadDelegate` is dropped
         self.track_ver.inc();
     }
 }
 
+/// #[RaftstoreCommon]
+pub trait ReadExecutorProvider<E>: Send + Clone + 'static
+where
+    E: KvEngine,
+{
+    type Delegate: ReadExecutor<E>;
+
+    fn store_id(&self) -> Option<u64>;
+
+    fn get_delegate_and_len(&self, region_id: u64) -> (usize, Option<Self::Delegate>);
+}
+
+#[derive(Clone)]
+pub struct StoreMetaDelegate<E>
+where
+    E: KvEngine,
+{
+    store_meta: Arc<Mutex<StoreMeta>>,
+    kv_engine: E,
+}
+
+impl<E> StoreMetaDelegate<E>
+where
+    E: KvEngine,
+{
+    pub fn new(store_meta: Arc<Mutex<StoreMeta>>, kv_engine: E) -> Self {
+        StoreMetaDelegate {
+            store_meta,
+            kv_engine,
+        }
+    }
+}
+
+impl<E> Deref for StoreMetaDelegate<E>
```

Comment:
I don't feel like this is particularly necessary. It it's only used between L272-L296, let's remove it.

Reply:
Done.